### DISCUSSION TITLE
讨论多档案切换相关 | Discuss about multi jsonrpc profile

### DISCUSSION THREAD

**Discussion by daiaji** [jc3213/download_with_aria2#22] [Show and tell]
完事了。@jc3213 
#20 
理论上是由于两次重装系统，进度缓了两天。
宝贵的假期啊。

**Comment by jc3213** [comment]
你试试看多档案切换，至少设置里没问题了。就是右键菜单你可能要等等……我到是想到了一个比较暴力的解决方案，但是这个方案有点容易引发错误……

**Reply by daiaji** [reply]
我不是chrome用户啊……

**Reply by jc3213** [reply]
火狐也支持了啊，后面我添加了，你没看新Issue么

**Reply by daiaji** [reply]
我的界面里找不到切换。

**Reply by jc3213** [reply]
你确定是从issue里的连接下载的multi_profile分支的源码么？主分支的源码啥都没的

**Comment by jc3213** [comment]
更新了说明 https://github.com/jc3213/download_with_aria2/issues/18

**Reply by jc3213** [reply]
@ivysrono 来帮忙测试一下吧。

**Comment by jc3213** [comment]
我已经把aria2.js用class重写了。这样适配多profile的时候也更容易了。回头再研究怎么操作

**Comment by jc3213** [comment]
@daiaji 多档案同步到最新的3.12.1了，现在在设置切换`Profile`后，点击`Expert`打开对应Profile的JSON-RPC的设置了，而之前的版本必须设置成为`Default`后才能打开。回头弄一下`Popup`。然后再说`Background`

**Reply by jc3213** [reply]
现在`Popup`也支持了，麻烦测试一下吧，`Background`再说

**Reply by daiaji** [reply]
wss似乎失效了，添在默认似乎也没用。

**Reply by jc3213** [reply]
我这里测试正常

![ws](https://user-images.githubusercontent.com/8744936/153712155-fcc58a18-fbdd-4e7b-825f-710cdf2b018f.png)
![popup](https://user-images.githubusercontent.com/8744936/153712160-4121aed4-d0f2-403e-8dd3-b1aabbba633a.png)

**Reply by daiaji** [reply]
{"id":"","jsonrpc":2,"method":"aria2.addUri","params":["token:token:445445454",["https://download.manjaro.org/kde/21.2.2/manjaro-kde-21.2.2-220123-linux515.iso"],{"out":"","header":["Cookie:","Referer: https://manjaro.org/downloads/official/kde/","User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.71 Safari/537.36"],"all-proxy":""}]}

**Reply by daiaji** [reply]
似乎是旧配置的问题。

**Reply by jc3213** [reply]
主要错误是`params":["token:token:445445454"` 这里多了`token:token:`

这个是因为3.12.1改了`Secret Token`的储存，因为使用class后可以只记录绝对值，然后在initiate的时候输入`'token:' + xxxx`，也可以保证这个不会多次计算这里。

然后旧的`secret_token`是储存`token:xxxx`，新的只需要储存`xxxx`，这里再更新日志里有提到，就是为了适配多档案。麻烦你改一下设置吧。

**Reply by jc3213** [reply]
我自己测试`popup`跟`options`的切换倒是没问题了。右键菜单以后找机会添加

**Reply by daiaji** [reply]
改了配置，暂时没什么特别值得说的问题。

**Reply by jc3213** [reply]
options跟popup没问题的话我找机会弄右键菜单。抓取的话还是保持默认为主。否则每次都要弹出窗口让你确认。

**Comment by jc3213** [comment]
菜单做出来了，但是功能没做，然后发现`notifications`相关的break了。找机会修复
![multi profile context menu](https://user-images.githubusercontent.com/8744936/153734887-88a8cc72-4261-46f9-92d6-4e95eed1a03e.png)

**Reply by jc3213** [reply]
右键菜单也能正常分工了，通知也修复了。
![multi profile working](https://user-images.githubusercontent.com/8744936/153735059-51922e8b-6776-4a00-bc8d-731e2b314fe0.png)

**Reply by daiaji** [reply]
我暂时没空，你把sh脚本丢到shell脚本格式化网站格式化一下吧，现在的对齐看起来怪瘆人的。

**Reply by jc3213** [reply]
已经弄好了，现在就看你们的反馈了。

------
- 右键菜单下载，并且把抓取用的JSONRPC修改成目标配置
- 右键菜单下载，下载过程中切换配置，下载完成后切回来 [**目前**]
------
- 一个主菜单，向后打开N个子菜单 [**目前**]
-  N个主菜单
------

我想知道这两个方案，不知道你觉得哪个好，等你实际使用后的反馈吧。包括右键菜单的形式，是多个菜单还是一个主菜单N个子菜单

**Reply by jc3213** [reply]
同步到3.12.2，然后移除了`confirm`确认框，因为Chromium那边有问题，稍微清理修复了下逻辑。等待深度测试

**Reply by daiaji** [reply]
多个子菜单吧

**Reply by daiaji** [reply]
firefox里没看见右键菜单。

**Reply by jc3213** [reply]
能给我看下错误么？我照着chromium的改的。理论没问题啊？

**Reply by daiaji** [reply]
Uncaught (in promise) ReferenceError: aria2RPCStatus is not defined
    aria2StartUp moz-extension://1f287d73-7fd8-4b4c-b854-9c1e00baa186/background.js:57
    <anonymous> moz-extension://1f287d73-7fd8-4b4c-b854-9c1e00baa186/background.js:9
[background.js:57:5](moz-extension://1f287d73-7fd8-4b4c-b854-9c1e00baa186/background.js)
***
https://archlinux.org/releng/releases/2022.02.01/torrent/
https://archlinux.org/download/
[background.js:42:17](moz-extension://1f287d73-7fd8-4b4c-b854-9c1e00baa186/background.js)
archlinux-2022.02.01-x86_64.iso.torrent [background.js:44:17](moz-extension://1f287d73-7fd8-4b4c-b854-9c1e00baa186/background.js)

**Reply by jc3213** [reply]
发现错误了，是换class以后火狐版本的indicator代码忘了改了……因为`3.12.0`抓取啥的都正常我也没多测试。最新的这个应该好了。顺便抓到一个BUG，保存也会修改Default部分代码的问题。

**Reply by daiaji** [reply]
右键菜单看见了,姑且用着没问题.Popup好像也正常显示了.

**Reply by jc3213** [reply]
有啥问题再反馈吧。我自己测试精力，能力有限。

**Comment by jc3213** [comment]
有个问题，多档案的运行状态指示器，因为运行在后台……导致切换没有那么容易……是否有必要把菜单改成
- 通过 Aria2 下载
- 切换档案 =>
    - 档案1
    - 档案2

这样做下载的时候会麻烦一点……但是可以灵活利用指示器……或者让Popup的下拉菜单可以直接切换档案？

因为感觉补这么做，指示器在多用户档案中似乎有点浮云……

**Comment by jc3213** [comment]
有人帮忙测试一下最新的代码么，我把`referer`跟`user-agent`从`header`里分离出来了，原因是参考了某个提取百度网盘的脚本，那个脚本得到的结果如下，因此我修改了原版跟多档案版的代码，我这边暂时没机会测试Firefox版本，只能测试Chrome版本貌似没什么问题？
![Baidu](https://user-images.githubusercontent.com/8744936/156871052-67bd7b7d-9aa6-44e1-a3fe-2efa179f7c48.png)

**Comment by jc3213** [comment]
重写了`Options`页面及逻辑，希望能给多档案提供一定的帮助……目前大概雏形有了，但是UI方面还有待优化。

**Comment by daiaji** [comment]
扬了？

**Reply by jc3213** [reply]
暂时还没新思路，我现在想改写options

**Reply by jc3213** [reply]
options改写完了，想休息一阵子了……前提是没有影响使用的问题。

再就是，我现在依然只能像最之前做出来的版本那样的思路去做多档案，感觉特别蠢？

**Reply by daiaji** [reply]
不方便是肯定的，不能绑定热键切换的话那也没什么办法，要不就是下载时加个弹窗用来切换档案。

**Reply by jc3213** [reply]
所以我制作了新的下载窗口……

**Reply by daiaji** [reply]
姑且先看看吧。

**Reply by jc3213** [reply]
右键菜单可以沿用那个思路，抓取可以改成弹出窗口选择，但是管理器这边就始终不合适。我还想弄一个按钮快速添加抓取当前页面的下载。

甚至学习其他扩展那样弄一个识别所有文件的下载（这个要用前台命令）

以及快捷方式，Hotkey + 点击链接 = 直接用aria2下载。当然这个也要前台命令

**Reply by daiaji** [reply]
发现所有的可下载内容确实算是实用功能。

**Reply by daiaji** [reply]
基本上以前的下载代理或者下载管理器实施的切换下载器方案好像就只有这些了，反正我记得另外那个集成ariang的aria代理就是用的这些交互方式。

**Reply by daiaji** [reply]
idm还是啥的，有个按住快捷键再点下载就不调用idm下载的功能，有些下载要验证乱七八糟的信息，用idm之类的下载器反而下载不了,然后本扩展注册浏览器里的磁力连接打开方式了吗？

**Reply by daiaji** [reply]
点击磁力连接下载应该用得着。

**Reply by jc3213** [reply]
> idm还是啥的，有个按住快捷键再点下载就不调用idm下载的功能，有些下载要验证乱七八糟的信息，用idm之类的下载器反而下载不了,然后本扩展注册浏览器里的磁力连接打开方式了吗？

这个我做了个功能，就是不转发请求头，但是目前还不能智能化，比如通过快捷键……然后关于注册磁力链接的打开方式，这里是知识盲区了……回头补一补，如果有资料的话希望提供一下

**Reply by jc3213** [reply]
> 点击磁力连接下载应该用得着。

这个理论上能做，但是必须跟快捷键+点击下载一起，毕竟监听的同一个事件。现在最大的问题是前后台交互，虽然我通过改写`New Download Window`已经预留了接口就是了

**Reply by jc3213** [reply]
> 发现所有的可下载内容确实算是实用功能。

最终问题还是前后台交互，这里我要看更多的资料。虽然知道是通过chrome.runtime.onMessage以及chrome.tabs.sendMessage或chrome.runtime.sendMessage操作。但是具体交互逻辑，界面生成啥的还需要研究，加上多用户档案，需要做的事情太多了，我经历又有限，属于边写边学，发现新东西或者能优化性能我就改写现有功能。因此进度会很慢很慢。

**Reply by daiaji** [reply]
> 然后关于注册磁力链接的打开方式，这里是知识盲区了……回头补一补，如果有资料的话希望提供一下

https://github.com/boypt/simple-torrent
这个项目的前端页面，会注册磁力连接，每次点开磁力链接的时候都会提示，前提大概是你访问过他的前端页面。

**Reply by jc3213** [reply]
用的navigator.registerProtocolHandler，但是这种需要远程服务器或者自建服务器去操作，前台链接并不能直接访问后台的extensions://id/handler，因此需要一个web://host/api?handler。基本上无解，只能作为Hotkey+左键菜单的一个补充。这里我还在想怎么弄。毕竟mv2，mv3某些东西实现不一样，单纯写content.js的话感觉后面会很复杂……再说吧

**Comment by jc3213** [comment]
@daiaji 试试看这个 [快捷Ctrl+Alt下载超链接或保存图片.zip](https://github.com/jc3213/download_with_aria2/files/10141664/Ctrl%2BAlt.zip) 覆盖到最新的1607分支下面就行了，我自己稍微测试了下，还行。pixiv的话event.preventDefault()不起作用但是还是能下载的（虽然不是100%触发）

**Reply by daiaji** [reply]
> @daiaji 试试看这个 [快捷Ctrl+Alt下载超链接或保存图片.zip](https://github.com/jc3213/download_with_aria2/files/10141664/Ctrl%2BAlt.zip) 覆盖到最新的1607分支下面就行了，我自己稍微测试了下，还行。pixiv的话event.preventDefault()不起作用但是还是能下载的（虽然不是100%触发）

直接覆盖到扩展包里？
按了没反应的样子。

**Reply by jc3213** [reply]
重新安装扩展并且刷新浏览的页面后才会生效

**Reply by jc3213** [reply]
我突然有了个新思路，就是content只做监听`message`事件，然后通过特定格式发送信息就能触发通过扩展下载。这样方便接入Tampermonkey，而且目前的这个**shortcut.js**，貌似问题不小……我看后台各种错误提示（很多界面无法工作的样子）

我自己的测试如下
![1](https://user-images.githubusercontent.com/8744936/205441149-bf01eee1-db4b-4443-adef-1e497a108d40.png)
![2](https://user-images.githubusercontent.com/8744936/205441152-29484047-a721-4508-b906-a7a35e6b6957.png)
![3](https://user-images.githubusercontent.com/8744936/205441153-18f10e5d-0c45-4487-a146-3bc6df4e0c70.png)
![4](https://user-images.githubusercontent.com/8744936/205441154-c4d2098e-4bda-4c63-9784-8efc0f6f3c38.jpg)

**Reply by ivysrono** [reply]
接入Tampermonkey是指只支持TM还是Greasemonkey和Violentmonkey这些脚本管理器都支持

**Reply by jc3213** [reply]
哪怕你在浏览器控制台输入，只要符合（目前我自己还在测试中的最新版本是） `window.postMessage( {aria2c: 'Download With Aria2', message: { url, referer, hostname, options} } )` 就行了，当然这个要等下一个正式版敲定以后再说。我还只是写着自己测试玩。

**Reply by jc3213** [reply]
对了，你们谁能提供一个，IDM或者FDM的扩展供我研究一下。我想知道快捷键抓取下载怎么实现的。我的方案总觉得很蠢

**Reply by jc3213** [reply]
@ivysrono 4.1.0.1612不用管，本来是想清理mv3里面多余的代码的，结果复制粘贴的时候manifest附带了后续版本的content-scripts，改来改去才修复。我吐了，以后除了核心版本，单独更新还是单独提出来吧。
